### PR TITLE
[ISSUE-26] Ensure that Lambdas are supported in ASM based transformer

### PR DESCRIPTION
**Pull request by scottmarlow**
Fix compile time support for Lambdas for [issue/26](https://github.com/wildfly-extras/batavia/issues/26)

### CODE REVIEW

File: impl/asm-based-transformer/src/main/java/org/wildfly/transformer/asm/TransformerImpl.java [hunk 1]
```diff
@@ -226,6 +228,86 @@ public void visitMethodInsn(int opcode, String owner, String name, String desc,
...
+                        /** bootstrapMethodArguments can be Integer, Float,Long, Double,String, Type, Handle, ConstantDynamic value. 
+                         * ConstantDynamic can modify content of array at runtime.
+                         */
+                        Object[] copyBootstrapMethodArguments = null;
+                        for(int looper = 0; looper < bootstrapMethodArguments.length; looper++) {
+                            Object argument = bootstrapMethodArguments[looper];
+                            if (argument instanceof Type) {
+                                Type type = (Type) argument;
+                                String oldDesc = type.getDescriptor();
+                                String updatedDesc = replaceJavaXwithJakarta(type.getDescriptor());
```
Comment (ropalka) [inline impl/asm-based-transformer/src/main/java/org/wildfly/transformer/asm/TransformerImpl.java:263]: We should do javax -> jakarta replacement magic here if, and only if 
type.getSort() returns ARRAY, OBJECT or METHOD int types.

For ARRAY types then we should:
 * check dimensions
 * replace descriptor (if necessary)
 * inspect and potentially replace all elements of this array type // see Type.getElementType()
 * inspect and potentially replace its internal name // see Type.getInternalName()

For METHOD types we should:
 * inspect and potentially replace all arguments of this method type // see Type.getArgumentTypes()
 * inspect and potentially replace its return type // see Type.getReturnType()

For OBJECT types we should:
 * inspect and potentially replace its internal name // see Type.getInternalName()
 * inspect and potentially replace its name // see Type.getClassName()

File: impl/asm-based-transformer/src/main/java/org/wildfly/transformer/asm/TransformerImpl.java [hunk 2]
```diff
@@ -226,6 +228,86 @@ public void visitMethodInsn(int opcode, String owner, String name, String desc,
...
+                                    if (copyBootstrapMethodArguments == null) {
+                                        copyBootstrapMethodArguments = cloneBootstrapMethodArguments(bootstrapMethodArguments);
+                                    }
+                                    copyBootstrapMethodArguments[looper] = Type.getMethodType(updatedDesc);
+                                }
+                            } else if (argument instanceof Handle) {  // reference to a field or method
+                                Handle handle = (Handle)argument;
+                                String origDesc = handle.getDesc();
+                                String updatedDesc = replaceJavaXwithJakarta(handle.getDesc());
+                                if (!origDesc.equals(updatedDesc)) {  // if we are changing
```
Comment (ropalka) [inline impl/asm-based-transformer/src/main/java/org/wildfly/transformer/asm/TransformerImpl.java:275]: Also handle.getOwner() should be inspected and potentially replaced if mapping applies to it.

File: impl/asm-based-transformer/src/main/java/org/wildfly/transformer/asm/TransformerImpl.java [hunk 3]
```diff
@@ -226,6 +228,86 @@ public void visitMethodInsn(int opcode, String owner, String name, String desc,
...
+                                    if (copyBootstrapMethodArguments == null) {
+                                        copyBootstrapMethodArguments = cloneBootstrapMethodArguments(bootstrapMethodArguments);
+                                    }
+                                    copyBootstrapMethodArguments[looper] = handle;
+                                }
+                                
+                            } else if( argument instanceof ConstantDynamic) {
+                                // TODO: runtime handling of ConstantDynamic)
+                                ConstantDynamic constantDynamic = (ConstantDynamic)argument;
+                                throw new IllegalStateException("ConstantDynamic is not handled " +constantDynamic.toString());
```
Comment (ropalka) [inline impl/asm-based-transformer/src/main/java/org/wildfly/transformer/asm/TransformerImpl.java:293]: I see https://github.com/wildfly-extras/batavia/issues/28 was created to cover that.

**Comment by scottmarlow** [comment]
> For METHOD types we should:
>     replace descriptor (if necessary)
>     inspect and potentially replace all arguments of this method type // see Type.getArgumentTypes()
>     inspect and potentially replace its return type // see Type.getReturnType()
 
When [iterating through bootstrapMethodArguments, for METHOD types, the Type.getReturnType throws:](https://github.com/scottmarlow/batavia/blob/lambdasdynamic/impl/asm-based-transformer/src/main/java/org/wildfly/transformer/asm/TransformerImpl.java#L291)
```
Exception in thread "main" java.lang.StringIndexOutOfBoundsException: String index out of range: 18
        at java.lang.String.charAt(String.java:658)
        at org.objectweb.asm.Type.getReturnTypeOffset(Type.java:390)
        at org.objectweb.asm.Type.getReturnType(Type.java:367)
        at org.objectweb.asm.Type.getReturnType(Type.java:356)
        at org.wildfly.transformer.asm.TransformerImpl$1$2.visitInvokeDynamicInsn(TransformerImpl.java:291)
        at org.objectweb.asm.ClassReader.readCode(ClassReader.java:2240)
        at org.objectweb.asm.ClassReader.readMethod(ClassReader.java:1283)
        at org.objectweb.asm.ClassReader.accept(ClassReader.java:688)
        at org.objectweb.asm.ClassReader.accept(ClassReader.java:400)
        at org.wildfly.transformer.asm.TransformerImpl.transform(TransformerImpl.java:76)
        at org.wildfly.transformer.asm.TransformerImpl.transform(TransformerImpl.java:504)
        at org.wildfly.transformer.tool.shared.Common.transformJarFile(Common.java:89)
        at org.wildfly.transformer.tool.cmdline.Main.main(Main.java:45)
```

**Comment by scottmarlow** [comment]
[test1.zip](https://github.com/wildfly-extras/batavia/files/4576443/test1.zip) contains a test jar

**Comment by scottmarlow** [comment]
Perhaps we need to check if the method is void.

**Comment by scottmarlow** [comment]
I see, the method parameter is of type Object, so that is why there is no return value type.

**Comment by scottmarlow** [comment]
I think that I see how we should handle dynamic Lambdas, one obstacle is that we haven't started a proper test suite, that could hit the various cases.  I'm thinking of breaking off the dynamic support, as a todo and add testsuite as a todo as well.  I would like to start working on the reflection case again, since that is a blocker currently (with Infinispan).

@ropalka agreed?

**Comment by scottmarlow** [comment]
replaced by [pull#33](https://github.com/wildfly-extras/batavia/pull/33)

**Comment by ropalka** [comment]
I agree @scottmarlow

**Pull request closed without merging**